### DISCUSSION TITLE
How do I share code between Supabase Edge functions (deno) and frontend(pnpm)?

### DISCUSSION THREAD

**Discussion by jerzakm** [supabase/supabase#12148] [Questions]
I'm a newbie working on a Sveltekit + supabase project. I use a lot of edge functions and I found myself rewriting a lot of the same utility functions for both. 
I'll appreciate any guides/examples of how to set up a mono repo like this:

Packages:
- webapp (pnpm, sveltekit)
- supabase (/functions - deno)
- utility (can be imported both by sveltekit and deno)

**Comment by jerzakm** [comment]
I have found no good solutions to this so far. 

Tldr: Best practices for sharing util functions between Deno edge functions and frontend sveltekit. Is my best bet waiting until Deno deploy supports 'npm:packagename' imports and then importing it like this?

**Comment by jsonMartin** [comment]
I'm also trying to figure this out

**Reply by jerzakm** [reply]
> I'm also trying to figure this out

I got a pretty good answer on discord. Tldr; mapping imports. It's pretty involved but I'll try it out soon.
https://discord.com/channels/839993398554656828/1076129210558599188/1076331509918728274

Quote:

Not a guide/example, but I found symlinks don't play nice, so I'm copying my shared code into the supabase/functions directory with a package.json script i.e.

"functions:prepare": "rm -rf ./supabase/functions/_* && cp -rf ./app/database/ ./supabase/functions/_database && cp -rf ./app/integrations/ ./supabase/functions/_integrations && cp -rf ./app/modules/ ./supabase/functions/_modules && cp -rf ./app/utils/ ./supabase/functions/_utils",

That'll get your code into a location that it'll be loaded into the Deno container. The next step is to make sure your imports work. You'll need to provide a Deno import_map.json. For the above, mine looks like:

{
  "imports": {
    "@supabase/supabase-js": "https://esm.sh/@supabase/supabase-js@2.7.0",
    "@supabase/supabase-js/": "https://esm.sh/@supabase/supabase-js@2.7.0/",
    "@supabase/postgrest-js": "https://esm.sh/@supabase/postgrest-js@1.3.0",
    "@supabase/postgrest-js/": "https://esm.sh/@supabase/postgrest-js@1.3.0/",
    "@trycourier/courier": "https://esm.sh/@trycourier/courier@4.1",
    "@trycourier/courier/": "https://esm.sh/@trycourier/courier@4.1/",
    "jose": "https://esm.sh/jose@4.11",
    "jose/": "https://esm.sh/jose@4.11/",
    "zod": "https://esm.sh/zod@3.20",
    "zod/": "https://esm.sh/x/zod@3.20/",
    "~/database/": "./functions/_database/",
    "./functions/_database/generated/client/runtime/index": "./functions/_database/generated/client/runtime/index.d.ts",
    "~/integrations/": "./functions/_integrations/",
    "~/modules/": "./functions/_modules/",
    "~/utils/webcrypto.node.server.ts": "./functions/_utils/webcrypto.deno.server.ts",
    "~/utils/": "./functions/_utils/"
  }
}


i.e. I'm importing code as ~/<something> in both my app and functions.

You'll note something else about my import map, I'm remapping NPM package named imports for zod, jose and of course supabase itself to Deno URL style imports.

Finally, you need to use .ts and .tsx file extensions everywhere. And here is where things get real tricky.

Until TS 5.0 (which isn't actually out yet)  .ts extensions weren't supported. However, TS 5.0 introduces compilerOptions  -> allowImportingTsExtensions which you'll need to set to true.  You'll want to make sure Node/package.json is using typescript 5.0.0-beta .

Depending on your IDE, there may be some more wrestling involved. For example, if you're using IntelliJ you'll absolutely need to be on the latest 2023.1 EAP for TS 5.0 support.

Finally, one additional gotcha. As much as I think Deno is interesting, their built-in integration of TS is a huge pain in the arse. In your node app (if you followed along) we've just configured it to use TS 5.0 beta. No such luck in Deno, you can't chose TS independently - especially since they've got their own Rust transpiling toolchain rather than TSC.

Anyway, the latest Deno is 1.30.3 but Supabase are using an outdated version - 1.28.0. I've submitted a PR to upgrade https://github.com/supabase/functions-relay/pull/30

What that means is that Supabase's edge runtime is currently 4.8.4. Not 5.0, and not even 4.9 i.e. you cannot currently use the satisfies keyword in any TS shared between node and your edge functions.

With this somewhat convoluted setup and those constraints in mind, sharing code does actually work 🙂
GitHub
[feat: Upgrade to Deno 1.30.3 by Benjamin-Dobell · Pull Request #30 ...](https://github.com/supabase/functions-relay/pull/30)
What kind of change does this PR introduce?
Upgraded to Deno 1.30.3. Main benefit is the upgrade to TypeScript 4.9.x, which introduces satisfies.
What is the current behavior?
Any attempts to deplo...
feat: Upgrade to Deno 1.30.3 by Benjamin-Dobell · Pull Request #30 ...
Oh, when running locally with supabase functions serve, you may run into issues with your import map. At least, that's what I'm currently fighting with. It works when I serve functions individually, but the new "serve all" functionality is not working as expected. I can't find the source for Supabase's edge-runtime container, so diagnosing is difficult.

**Reply by laktek** [reply]
> Oh, when running locally with supabase functions serve, you may run into issues with your import map. At least, that's what I'm currently fighting with. It works when I serve functions individually, but the new "serve all" functionality is not working as expected. I can't find the source for Supabase's edge-runtime container, so diagnosing is difficult.

@jerzakm This issue should be fixed from Supabase CLI 1.38.6+. You should be able to use an import map with "serve all".
